tmt/steps/execute/__init__.py

```diff
@@ -602,6 +609,107 @@ def handle_reboot(self) -> bool:
+            interaction with user.
+        :param timeout: if set, command would be interrupted, if still running,
+            after this many seconds.
+        :param log: a logging function to use for logging of command output. By
+            default, ``logger.debug`` is used.
+        :returns: command output.
+        """
+
+        # Write down process and let tmt kill it if tmt gets interrupted.
+        def _save_process(
+            command: Command, process: subprocess.Popen[bytes], logger: tmt.log.Logger
+        ) -> None:
+            """
+            Record process info in the invocation.
+
+            Called by :py:class:`Command` as the ``on_process_start``
+            callback.
+            """
+
+            with self.process_lock:
+                self.process = process
+
+                self.on_interrupt_callback_token = tmt.utils.signals.add_callback(
+                    self.terminate_process, logger=self.logger
+                )
+
+        def _reset_process(
+            command: Command,
+            process: subprocess.Popen[bytes],
+            output: CommandOutput,
+            logger: tmt.log.Logger,
+        ) -> None:
+            """
+            Reset process info in the invocation.
+
+            Called by :py:class:`Command` as the ``on_process_end``
+            callback.
+            """
+
+            with self.process_lock:
+                self.process = None
+
+                if self.on_interrupt_callback_token is not None:
+                    tmt.utils.signals.remove_callback(self.on_interrupt_callback_token)
+
+        with Stopwatch() as timer:
+            self.start_time = format_timestamp(timer.start_time)
+
+            try:
+                output = self.guest.execute(
+                    command,
+                    cwd=cwd,
+                    env=env,
+                    join=True,
+                    interactive=interactive,
+                    tty=self.test.tty,
+                    log=log,
+                    timeout=timeout,
+                    on_process_start=_save_process,
+                    on_process_end=_reset_process,
+                    test_session=True,
+                    friendly_command=str(self.test.test),
+                )
+
+                self.return_code = tmt.utils.ProcessExitCodes.SUCCESS
+
+            except tmt.utils.RunError as error:
+                output = error.output
+
+                self.return_code = error.returncode
+
+                if self.return_code == tmt.utils.ProcessExitCodes.TIMEOUT:
+                    self.logger.debug(f"Test duration '{self.test.duration}' exceeded.")
+
+                elif tmt.utils.ProcessExitCodes.is_pidfile(self.return_code):
+                    self.logger.warning('Test failed to manage its pidfile.')
+
+        self.end_time = format_timestamp(timer.end_time)
+        self.real_duration = format_duration(timer.duration)
```

Comment:
Shouldn't these two lines be indented under the `with Stopwatch()` statement?

Reply:
I don't think they can, as `Stopwatch.end_time` is set when leaving the context, i.e. oved under `with Stopwatch()`, it would not be set to anything useful. And `Stopwatch.duration` depends on `end_time`, same fate.

Reply:
Ah, I see, just read the `Stopwatch` code. Now it make sense. Thanks for clarification.